NodeApi/JSPropertyDescriptorListOfT.cs

```diff
@@ -0,0 +1,188 @@
+using System;
+using System.Collections.Generic;
+
+namespace NodeApi;
+
+public class JSPropertyDescriptorList<TDerived, TObject>
+  where TDerived : class, IJSObjectUnwrap<TObject>
+  where TObject : class
+{
+  public IList<JSPropertyDescriptor> Properties { get; } = new List<JSPropertyDescriptor>();
+
+  public JSPropertyDescriptorList() { }
+
+  public TDerived AddProperty(
+    string name,
+    JSValue value,
+    JSPropertyAttributes attributes = JSPropertyAttributes.DefaultProperty)
+  {
+    Properties.Add(new JSPropertyDescriptor(name, value, attributes));
+    return (TDerived)(object)this;
+  }
+
+  public TDerived AddProperty(
+    string name,
+    JSCallback? getter,
+    JSCallback? setter,
+    JSPropertyAttributes attributes = JSPropertyAttributes.DefaultProperty)
+  {
+    Properties.Add(new JSPropertyDescriptor(name, getter, setter, attributes));
+    return (TDerived)(object)this;
+  }
+
+  public TDerived AddProperty(
+    string name,
+    Func<TObject, JSValue>? getter,
+    Action<TObject, JSValue>? setter,
+    JSPropertyAttributes attributes = JSPropertyAttributes.DefaultProperty)
+  {
+    return AddProperty(
+      name,
+      getter == null ? null : args => (TDerived.Unwrap(args) is TObject obj) ? getter(obj) : JSValue.Undefined,
+      setter == null ? null : args =>
+      {
+        if (TDerived.Unwrap(args) is TObject obj)
+        {
+          setter(obj, args[0]);
+        }
+        return JSValue.Undefined;
+      },
+      attributes);
+  }
+
+  public TDerived AddProperty(
+    string name,
+    Func<JSValue>? getter,
+    Action<JSValue>? setter,
+    JSPropertyAttributes attributes = JSPropertyAttributes.DefaultProperty | JSPropertyAttributes.Static)
+  {
+    return AddProperty(
+      name,
+      getter == null ? null : args => getter(),
+      setter == null ? null : args =>
+      {
+        setter(args[0]);
+        return JSValue.Undefined;
+      },
+      attributes);
+  }
+
+  public TDerived AddMethod(
+    string name,
+    JSCallback callback,
+    JSPropertyAttributes attributes = JSPropertyAttributes.DefaultMethod)
+  {
+    Properties.Add(new JSPropertyDescriptor(name, callback, attributes));
+    return (TDerived)(object)this;
+  }
+
+  public TDerived AddMethod(
+    string name,
+    Func<TObject, Action> getMethod,
```

Comment:
I'm confused by the extra level of indirection for the method callbacks. Why wouldn't this parameter be just `Action<TObject, JSCallbackArgs>` or `Func<TObject, JSCallbackArgs, JSValue>` (or variations without args)? Is there some need for lazy initialization of the method callbacks?

Reply:
It is just a sketch for how we can implement methods with custom signatures. 
We can use generic arguments `T1`, `T2`, etc. instead of `JSValue` in these signatures.
In the end they all call the `AddMethod` with `JSCallback` that accepts `JSCallbackArgs` and returns `JSValue`.

We can bypass the indirection, but then we must require the generated user code to use the unsafe API directly.